rust/src/mime/smtp.rs

```diff
@@ -0,0 +1,599 @@
+            if mime::rs_equals_lowercase(&h.value, b"base64") {
+                ctx.encoding = MimeSmtpEncoding::Base64;
+            } else if mime::rs_equals_lowercase(&h.value, b"quoted-printable") {
+                ctx.encoding = MimeSmtpEncoding::QuotedPrintable;
+            }
+        }
+    }
+    for h in &ctx.headers[ctx.main_headers_nb..] {
+        if mime::rs_equals_lowercase(&h.name, b"content-type") {
+            if ctx.filename.is_empty() {
+                if let Some(value) =
+                    mime::mime_find_header_token(&h.value, b"name", &mut sections_values)
+                {
+                    ctx.filename.extend_from_slice(value);
+                    let mut newname = Vec::new();
+                    newname.extend_from_slice(value);
+                    ctx.attachments.push(newname);
+                    sections_values.clear();
+                }
+            }
+            if ctx.main_headers_nb == 0 {
+                if let Some(value) =
+                    mime::mime_find_header_token(&h.value, b"boundary", &mut sections_values)
+                {
+                    // start wih 2 additional hyphens
+                    ctx.boundary.push(b'-');
+                    ctx.boundary.push(b'-');
+                    ctx.boundary.extend_from_slice(value);
+                    sections_values.clear();
+                }
+            }
+            let ct = if let Some(x) = h.value.iter().position(|&x| x == b';') {
+                &h.value[..x]
+            } else {
+                &h.value
+            };
+            match ct {
+                b"text/plain" => {
+                    ctx.content_type = MimeSmtpContentType::PlainText;
+                }
+                b"text/html" => {
+                    ctx.content_type = MimeSmtpContentType::Html;
+                }
+                _ => {
+                    ctx.content_type = MimeSmtpContentType::Unknown;
+                }
+            }
+            break;
+        }
+    }
+}
+
+extern "C" {
+    // Defined in util-file.h
+    pub fn FileAppendData(
+        c: *mut FileContainer, data: *const c_uchar, data_len: u32,
+    ) -> std::os::raw::c_int;
+    // Defined in util-spm-bs.h
+    pub fn BasicSearchNocaseIndex(
+        data: *const c_uchar, data_len: u32, needle: *const c_uchar, needle_len: u16,
+    ) -> u32;
+}
+
+fn hex(i: u8) -> Option<u8> {
+    if (b'0'..=b'9').contains(&i) {
+        return Some(i - b'0');
+    }
+    if (b'A'..=b'F').contains(&i) {
+        return Some(i - b'A' + 10);
+    }
+    return None;
+}
+
+const SMTP_MIME_MAX_DECODED_LINE_LENGTH: usize = 8192;
+
+fn mime_smtp_finish_url(input: &[u8]) -> &[u8] {
+    if let Some(x) = input.iter().position(|&x| {
+        x == b' ' || x == b'"' || x == b'\'' || x == b'<' || x == b'>' || x == b']' || x == b'\t'
+    }) {
+        return &input[..x];
+    }
+    return input;
+}
+
+fn mime_smtp_extract_urls(urls: &mut Vec<Vec<u8>>, input_start: &[u8]) {
+    //TODOrust6 use mpm
```

Comment:
thoughts about this optimization ? (using a mpm)

Reply:
seems nice but can be done later?